crates/pathfinder/src/state/sync/l2.rs

```diff
@@ -185,7 +187,20 @@ async fn download_block(
     let result = sequencer.block(block_number.into()).await;
 
     match result {
-        Ok(block) => Ok(DownloadBlock::Block(Box::new(block))),
+        Ok(block) => {
+            // Check if block hash is correct.
+            let expected_block_hash = block.block_hash.unwrap();
+            let verify_hash = tokio::task::spawn_blocking(move || -> anyhow::Result<_> {
+                let verify_result = verify_block_hash(&block, chain, expected_block_hash)?;
+                Ok((block, verify_result))
+            });
+            let (block, verify_result) = verify_hash.await.context("Verify block hash")??;
+            if verify_result == VerifyResult::Mismatch {
+                let block_number = block.block_number.unwrap();
+                tracing::warn!(?block_number, block_hash = ?expected_block_hash, "Block hash mismatch");
+            }
+            Ok(DownloadBlock::Block(Box::new(block)))
+        }
```

Comment:
This makes me think that the boxing should happen first, then moving the value and returning it from the spawned task would be pointer move.

Reply:
That's a good idea -- I moved the boxing to the top of the block. We can follow-up in the sequencer as a separate PR.